src/sse_events.rs

```diff
@@ -0,0 +1,154 @@
+use crate::{
+    console::{ConsoleOutput, SerializableEvent},
+    demo::{DemoEvent, DemoMessage, VoteCastEventWrapped, VoteRelatedEvent},
+    state::MACState,
+    web::broadcast_event,
+};
+use event_loop::{try_get, Handled, HandlerStruct, Is};
+use std::collections::HashMap;
+
+/// This struct will house the relevant vars used by various message handlers when broadcasting
+/// Encapsulates a set of functions to invoke the `web::broadcast_event` function with the
+/// appropriately serialised JSON messages.
+pub struct BroadcastableEvent {
+    /// Used when handling certain `DemoMessage` messages (I.e. `VoteStarted`)
+    votes: HashMap<u32, Vec<String>>,
+}
+
+impl BroadcastableEvent {
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            votes: HashMap::new(),
+        }
+    }
+}
+
+impl Default for BroadcastableEvent {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// We 'broadcast' events to subscribers using SSE (Server Sent Events) publishing to push events to
+/// clients without them having to poll us. This is essentially a reverse API, where we expose the
+/// endpoint, but data is shipped when _we_ want and the clients have to respond.
+///
+/// See `broadcast_event` in `crate::web` for more info
+impl<IM, OM> HandlerStruct<MACState, IM, OM> for BroadcastableEvent
+where
+    IM: Is<DemoMessage> + Is<ConsoleOutput>,
+{
+    /// `IM` can be a message that is either a `DemoMessage` or a `ConsoleOutput` message
+    /// These are then dispatched to the appropriate discrete message handler, which take the
+    /// direct message (i.e. not wrapped in Is<> or as a generic), then return `Option<String>` for
+    /// the message serialisation.
+    /// If a `String` was returned, we broadcast that message to all subscribers.
+    #[allow(clippy::cognitive_complexity)]
+    fn handle_message(&mut self, state: &MACState, message: &IM) -> Option<Handled<OM>> {
+        let event_json = if try_get::<DemoMessage>(message).is_some() {
+            let demo_msg = try_get::<DemoMessage>(message)?;
+            self.handle_demo_message(state, demo_msg)
+        } else if try_get::<ConsoleOutput>(message).is_some() {
+            let con_msg = try_get::<ConsoleOutput>(message)?;
+            self.handle_console_message(state, con_msg)
+        } else {
+            None
+        };
+        if let Some(json) = event_json {
+            Handled::<OM>::future(async move {
+                broadcast_event(json).await;
+                None
+            })
+        } else {
+            None
+        }
```

Comment:
I'd personally opt for 
```rust
    ...
}

None
```
But doesn't really matter.

Reply:
in that case, i'd have to make the statement inside the if body be a `return`; 
```rs
        if let Some(json) = event_json {
            Handled::<OM>::future(async move {
                broadcast_event(json).await;
                None
            })
        } else {
            None
        }
```
vs
```rs
        if let Some(json) = event_json {
            return Handled::<OM>::future(async move {
                broadcast_event(json).await;
                None
            });
        }
        None
```

Reply:
Yeah true. I'm not really bothered either way, was just what I would have done. Up to you.